Add new module for elasticache parameter group.

akurniawan opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
elasticache_parameter_group module

##### ANSIBLE VERSION
```
2.2.0
```

##### SUMMARY
This is new module to create parameter group for elasticache in AWS

Commit:
Add new module for elasticache parameter group.

Commit:
Add Sha-bang

Commit:
fix iteritems

Commit:
add GPLv3 license

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,239 @@
+        response, changed = upsert_elasticache_pg(
+            es_client, existing_pg, module)
+
+    if changed:
```
you can just `module.exit_json(changed=changed, parameter_groups=response)`

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,239 @@
+
+def get_es_client(params):
+    """Get elasticache client"""
+    session = boto3.Session(aws_access_key_id=params["aws_access_key"],
```
we should be using boto3_conn for all connections. For example, security_token is missing here. boto3_conn avoids you having to worry about that.

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,239 @@
+try:
+    import boto3
+    import botocore
+    HAS_BOTO = True
```
should be HAS_BOTO3

zimbatm commented:
Please remove me from the bot mentions. The linked help page doesn't give me a way to do it myself.

Commit:
- update the way of gathering ec2 connection using boto3_conn
- update HAS_BOTO to HAS_BOTO3
- update logic from exiting module

akurniawan commented:
Hi @zimbatm, how exactly can I do that?

Commit:
update over-indented lines

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,237 @@
+description:
+  - Manage elasticache parameter group
+  - Has the capabilities to create/destroy parameter group
+version_added: "2.2"
```
Should be 2.3

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,237 @@
+
+
+def convert_params_format(params):
+    result = []
```
This method can be simplified to:

```
return [{"ParameterName": str(k), "ParameterValue": str(v)} for (k,v) in params.items]
```

willthames replied:
This is a suggestion, not a blocker.

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,237 @@
+
+    module.exit_json(changed=changed, parameter_groups=response)
+
+from ansible.module_utils.basic import *
```
Should be changed to new style module imports nearer the top of the file:

```
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.ec2 import ec2_argument_spec, get_aws_connection_info, boto3_conn
```

willthames replied:
This is still outstanding

akurniawan replied:
should I put the imports to the very beginning of the file or above DOCUMENTATION variable?

willthames replied:
I put mine after the comments just above the documentation variable as then it passes pep8.

Commit:
- update ansible version
- change convert_params_format to one liner
- move import AnsibleModule and helper functions to nearer the top of the file

Commit:
update imports placement

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,231 @@
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+try:
```
flake8 will moan about this - and apologies for not being clearer. Leave this section where it was, and then everything should be fine. One thing to note while I'm in the area,but it's not a blocker, is that flake8 will complain less if you add HAS_BOTO3 to the list of imports and don't import boto3 (in the `except ImportError` part, just `pass`, as you'll handle that  using HAS_BOTO3 later)

akurniawan replied:
I don't quite follow this part 
> add HAS_BOTO3 to the list of imports and don't import boto3. 

Is this what you mean?
try:
    import botocore
    HAS_BOTO3 = True
except ImportError:
    pass

How can I tell if boto3 is really installed?

willthames replied:
```
from ansible.module_utils.ec2 import HAS_BOTO3

try:
    import botocore
except ImportError:
    pass  # will be caught by imported HAS_BOTO3
```

willthames commented:
I'm sorry, I've given bad advice - ansible/proposals#50 now means that, in contradiction to what standard flake8 says, the imports should be after all the documentation. This has been a little bit of a moving target during my reviews of this PR, but I still apologise for the misinformation. Should be an easy fix though.

Commit:
move import to line after documentation

Commit:
add ANSIBLE_METADATA

s-hertel commented:
Closing because pr #21023 was merged. If you think there is missing functionality, please let me know.
